In [9]:
# ---------------------------------------------Задание №1 -----------------------------------------------
import pandas as pd

# Загрузим файл и посмотрим на данные
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Машинное обучение/HR.csv')
data.head()


satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler

# Кодируем категориальные переменные
data_encoded = data.copy()
le_department = LabelEncoder()
le_salary = LabelEncoder()

data_encoded['department'] = le_department.fit_transform(data['department'])
data_encoded['salary'] = le_salary.fit_transform(data['salary'])

# Масштабируем числовые переменные для моделей
scaler = StandardScaler()
scaled_columns = ['satisfaction_level', 'last_evaluation', 'number_project',
                  'average_montly_hours', 'time_spend_company']
data_encoded[scaled_columns] = scaler.fit_transform(data_encoded[scaled_columns])

# Разделение данных для задачи 1
X_task1 = data_encoded.drop('left', axis=1)
y_task1 = data_encoded['left']
X_train_task1, X_test_task1, y_train_task1, y_test_task1 = train_test_split(X_task1, y_task1, test_size=0.2, random_state=42)

X_task1.head()


satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0           -0.936495        -1.087275       -1.462863             -0.882040   
1            0.752814         0.840707        0.971113              1.220423   
2           -2.022479         0.957554        2.593763              1.420657   
3            0.431041         0.899131        0.971113              0.439508   
4           -0.976716        -1.145699       -1.462863             -0.841993   

   time_spend_company  Work_accident  promotion_last_5years  department  \
0           -0.341235              0                      0           7   
1            1.713436              0                      0           7   
2            0.343655              0                      0           7   
3            1.028546              0                      0           7   
4           -0.341235              0                      0           7   

   salary  
0       1  
1       2  
2       2  
3       1  
4       1

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.metrics import classification_report, accuracy_score

# Логистическая регрессия
log_reg = LogisticRegression(random_state=42)
log_reg.fit(X_train_task1, y_train_task1)
y_pred_log = log_reg.predict(X_test_task1)
acc_log = accuracy_score(y_test_task1, y_pred_log)

# Случайный лес
rf = RandomForestClassifier(random_state=42, n_estimators=100)
rf.fit(X_train_task1, y_train_task1)
y_pred_rf = rf.predict(X_test_task1)
acc_rf = accuracy_score(y_test_task1, y_pred_rf)

# Градиентный бустинг
gb = GradientBoostingClassifier(random_state=42)
gb.fit(X_train_task1, y_train_task1)
y_pred_gb = gb.predict(X_test_task1)
acc_gb = accuracy_score(y_test_task1, y_pred_gb)

# Ансамблирование
voting_clf = VotingClassifier(estimators=[
    ('log_reg', log_reg),
    ('rf', rf),
    ('gb', gb)
], voting='hard')

voting_clf.fit(X_train_task1, y_train_task1)
y_pred_voting = voting_clf.predict(X_test_task1)
acc_voting = accuracy_score(y_test_task1, y_pred_voting)

# Вывод результатов
{
    "Logistic Regression": acc_log,
    "Random Forest": acc_rf,
    "Gradient Boosting": acc_gb,
    "Voting Classifier": acc_voting
}


{'Logistic Regression': 0.7576666666666667,
 'Random Forest': 0.9883333333333333,
 'Gradient Boosting': 0.9736666666666667,
 'Voting Classifier': 0.9783333333333334}

Результаты точности моделей для предсказания текучести кадров:

- Логистическая регрессия: 75.77%
- Случайный лес: 98.83%
- Градиентный бустинг: 97.37%
- Ансамблирование: 97.83%
<p> Увы, но точность >99% нигде не наблюдается. Зато случайный лес дал наибольшую.


In [12]:
# ---------------------------------------------Задание №2 -----------------------------------------------